[ZIP 48] P2SH multisig (#1074)
zcash/zips: zips/zip-0048.md

Below here is proposed specification text:

Zcash wallets SHOULD use the BIP 383 [^bip-0383] `sortedmulti()` script expression to
produce Zcash P2MS scripts, which would then be contained in the BIP 381 [^bip-0381]
`sh()` script expression to produce Zcash P2SH addresses.

The following BIP 388 [^bip-0388] descriptor template produces a 2-of-3 P2SH multisig
wallet compatible with this ZIP:

```
sh(sortedmulti(2,@0/**,@1/**,@2/**))
```

## Hierarchical Derivation

Given a set of participants that each have some seed material (e.g. BIP 39 [^bip-0039]
mnemonic phrases), and a BIP 380 [^bip-0380] output descriptor that encodes the desired
multisig policy, wallets need to produce a sequence of P2SH addresses that can be given
out for receiving funds, and scanned to detect incoming payments. This specification
adopts BIP 48 [^bip-0048] for this purpose, with the following modification:

- The only defined $\mathit{script\_type}$ is $\mathtt{0x133000}$ for “Zcash P2SH” as
  defined in this specification. The examples for Nested Segwit and Native Segwit are
  therefore not applicable.

## Transaction construction

TODO: Anything we need to specify here?

# Rationale

> We specify a likely-non-colliding $\mathit{script\_type}$ instead of using either `0` (the gap
> left in BIP 48) or `2` (the next unassigned contant in BIP 48), because both of those values

Reviewer: ```suggestion
We specify a likely-non-colliding $\mathit{script\_type}$ instead of using either $0'$ (the gap
left in BIP 48) or $2'$ (the next unassigned contant in BIP 48), because both of those values
```

BIP 48), because both of those values have in the past been proposed and rejected as
referring to P2SH. This means that there is no compatibility benefit to trying to use one
or the other, and the likelihood of a collision with some other use case is higher.
Instead we pick a value that is prefixed (in decimal) with the SLIP 44 [^slip-0044]
$\mathit{coin\_type}$ for Zcash mainnet, which is highly unlikely to organically occur in
another related specification.

By not using BIP 45, Zcash addresses will not be compatible with BIP 45 P2SH derivation
paths. We consider this acceptable because we explicitly do not want users that use the
same mnemonic phrase for both Bitcoin and Zcash to produce addresses that reuse the same
public keys.

By not using BIP 45 (with a $\mathit{cosigner\_index}$ path element), all cosigners produce
the same sequence of P2SH addresses. This is good for agreement on e.g. the ZIP 1016 [^zip-1016]
Key-Holder Organizations’ Mainnet address, but in general usage means that two cosigners
proposing transactions can race and use the same P2SH address twice. This could result in a
situation where different members of the signing set inadvertently give the same P2SH

> address to different counterparties, inadvertently linking the associated payment

Reviewer 2: ```suggestion
address to different counterparties, linking the associated payment
```

activity. We consider this tradeoff acceptable because it aligns with what hardware
wallets have done for years (implying that users are fine with the usability). We consider

> the risk acceptable because avoiding linkage between transparent addrs is hard in any

Reviewer 2: ```suggestion
the risk acceptable because avoiding linkage between transparent addresses is hard in any
```

case, and not something that Zcash actively attempts to avoid (the shielded protocols are

> for this use case).

Reviewer 2: ```suggestion
designed for this use case).
```
